lib/src/java/facter.cc

```diff
@@ -0,0 +1,169 @@
+#include <facter/facts/scalar_value.hpp>
+#include <facter/facts/array_value.hpp>
+#include <facter/facts/map_value.hpp>
+#include <facter/logging/logging.hpp>
+#include <facter/export.h>
+#include <boost/nowide/iostream.hpp>
+#include <string>
+#include <memory>
+
+using namespace std;
+using namespace facter::facts;
+using namespace facter::logging;
+
+static jclass object_class, long_class, double_class, boolean_class, hash_class;
+static jmethodID long_constructor, double_constructor, boolean_constructor, hash_constructor, hash_put;
+static std::unique_ptr<collection> facts_collection;
+
+static string to_string(JNIEnv* env, jstring str)
+{
+    if (!str) {
+        return {};
+    }
+
+    auto ptr = env->GetStringUTFChars(str, nullptr);
+    if (!ptr) {
+        return {};
+    }
+
+    auto size = env->GetStringUTFLength(str);
+    string result(ptr, size);
+
+    env->ReleaseStringUTFChars(str, ptr);
+    return result;
+}
+
+static jobject to_object(JNIEnv* env, value const* val)
+{
+    if (!val) {
+        return nullptr;
+    }
+    if (auto ptr = dynamic_cast<string_value const*>(val)) {
+        return env->NewStringUTF(ptr->value().c_str());
+    }
+    if (auto ptr = dynamic_cast<integer_value const*>(val)) {
+        return env->NewObject(long_class, long_constructor, static_cast<jlong>(ptr->value()));
+    }
+    if (auto ptr = dynamic_cast<boolean_value const*>(val)) {
+        return env->NewObject(boolean_class, boolean_constructor, static_cast<jboolean>(ptr->value()));
+    }
+    if (auto ptr = dynamic_cast<double_value const*>(val)) {
+        return env->NewObject(double_class, double_constructor, static_cast<jdouble>(ptr->value()));
+    }
+    if (auto ptr = dynamic_cast<array_value const*>(val)) {
+        auto array = env->NewObjectArray(ptr->size(), object_class, nullptr);
+
+        // Recurse on each element of the array
+        jsize index = 0;
+        ptr->each([&](value const* element) {
+            env->SetObjectArrayElement(array, index++, to_object(env, element));
+            return true;
+        });
+        return array;
+    }
+    if (auto ptr = dynamic_cast<map_value const*>(val)) {
+        auto hashmap = env->NewObject(hash_class, hash_constructor, static_cast<jint>(ptr->size()));
+
+        // Recurse on each element in the map
+        ptr->each([&](string const& name, value const* element) {
+            env->CallObjectMethod(hashmap, hash_put, env->NewStringUTF(name.c_str()), to_object(env, element));
+            return true;
+        });
+        return hashmap;
+    }
+    return nullptr;
+}
+
+extern "C" {
+    LIBFACTER_EXPORT jint JNI_OnLoad(JavaVM* vm, void* reserved)
+    {
+        JNIEnv* env;
+        if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) != JNI_OK) {
+            return JNI_ERR;
+        }
+        // For the classes, we need global refs
+        object_class = static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/lang/Object")));
```

Comment:
Generally `FindClass` can return `null` though I would never expect it to fail to load these classes.